test/test_model.py

```diff
@@ -3228,43 +3347,67 @@ def test_set_info(self):
             secret.set_info()  # no args provided
 
     def test_grant(self):
+        fake_script(self, 'relation-list', """echo '[]'""")
         fake_script(self, 'secret-grant', """exit 0""")
         fake_script(self, 'secret-info-get', """echo '{"z": {"label": "y", "revision": 7}}'""")
 
         secret = self.make_secret(id='x')
-        secret.grant(types.SimpleNamespace(id=123))
-        secret.grant(types.SimpleNamespace(id=234), unit=types.SimpleNamespace(name='app/0'))
+        backend = ops.model._ModelBackend('test', 'test', 'test')
+        meta = ops.CharmMeta()
+        cache = ops.model._ModelCache(meta, backend)
+        unit = ops.Unit('test', meta, backend, cache)
+        rel123 = ops.Relation('test', 123, True, unit, backend, cache)
+        rel234 = ops.Relation('test', 234, True, unit, backend, cache)
+        secret.grant(rel123)
+        unit = ops.Unit('app/0', meta, backend, cache)
+        secret.grant(rel234, unit=unit)
```

Comment:
Seeing we're probing into internals anyway, I wonder if we should just reuse the existing backend and cache objects?

```suggestion
        secret = self.make_secret(id='x')
        unit = ops.Unit('test', ops.CharmMeta(), self.model._backend, self.model._cache)
        rel123 = ops.Relation('test', 123, True, unit, self.model._backend, self.model._cache)
        rel234 = ops.Relation('test', 234, True, unit, self.model._backend, self.model._cache)
        secret.grant(rel123)
        unit = ops.Unit('app/0', ops.CharmMeta(), self.model._backend, self.model._cache)
        secret.grant(rel234, unit=unit)
```